Objective

The objective of this notebook is to clean the raw dataset and prepare it for further analysis and machine learning. During this phase, unnecessary columns and highly incomplete features are removed, and the dataset structure is improved for subsequent feature selection and model development.

In [2]:
# ==========================================
# MULE ACCOUNT DETECTION
# PHASE 2 : DATA CLEANING &
# MISSING VALUE TREATMENT
# ==========================================

import pandas as pd
import numpy as np
import os

# ------------------------------------------
# LOAD DATASET
# ------------------------------------------

df = pd.read_csv("../data/raw/DataSet.csv")

print("=" * 60)
print("ORIGINAL DATASET")
print("=" * 60)

print("Shape:", df.shape)

# ------------------------------------------
# REMOVE UNNECESSARY COLUMN
# ------------------------------------------

if "Unnamed: 0" in df.columns:
    df.drop(columns=["Unnamed: 0"], inplace=True)

print("\nAfter Removing Unnamed: 0")
print("Shape:", df.shape)

# ------------------------------------------
# MISSING VALUE ANALYSIS
# ------------------------------------------

missing_percent = (
    df.isnull().sum() / len(df)
) * 100

# ------------------------------------------
# FIND COLUMNS >80% MISSING
# ------------------------------------------

high_missing_cols = missing_percent[
    missing_percent > 80
].index

print("\nColumns >80% Missing:")
print(len(high_missing_cols))

# ------------------------------------------
# DROP HIGH MISSING COLUMNS
# ------------------------------------------

df.drop(
    columns=high_missing_cols,
    inplace=True
)

print("\nAfter Removing High Missing Columns")
print("Shape:", df.shape)

# ------------------------------------------
# TARGET COLUMN
# ------------------------------------------

target_col = "F3924"

# ------------------------------------------
# NUMERICAL FEATURES
# ------------------------------------------

numerical_cols = df.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

if target_col in numerical_cols:
    numerical_cols.remove(target_col)

# ------------------------------------------
# CATEGORICAL FEATURES
# ------------------------------------------

categorical_cols = df.select_dtypes(
    include=["object", "string"]
).columns.tolist()

print("\nNumerical Features:", len(numerical_cols))
print("Categorical Features:", len(categorical_cols))

# ------------------------------------------
# MISSING VALUES BEFORE TREATMENT
# ------------------------------------------

missing_before = df.isnull().sum().sum()

print("\nMissing Values Before Treatment:")
print(f"{missing_before:,}")

# ------------------------------------------
# FILL NUMERICAL FEATURES
# ------------------------------------------

print("\nFilling Numerical Features...")

for col in numerical_cols:
    median_value = df[col].median()
    df[col] = df[col].fillna(median_value)

# ------------------------------------------
# FILL CATEGORICAL FEATURES
# ------------------------------------------

print("Filling Categorical Features...")

for col in categorical_cols:
    df[col] = df[col].fillna("Unknown")

# ------------------------------------------
# VERIFY MISSING VALUES
# ------------------------------------------

missing_after = df.isnull().sum().sum()

print("\nMissing Values After Treatment:")
print(f"{missing_after:,}")

# ------------------------------------------
# TARGET DISTRIBUTION CHECK
# ------------------------------------------

print("\nTarget Distribution")

print(
    df[target_col]
    .value_counts()
)

# ------------------------------------------
# SAVE PREPROCESSED DATASET
# ------------------------------------------

os.makedirs(
    "../data/processed",
    exist_ok=True
)

df.to_csv(
    "../data/processed/preprocessed_data.csv",
    index=False
)

print("\nPreprocessed Dataset Saved")

# ------------------------------------------
# FINAL SUMMARY
# ------------------------------------------

print("\n" + "=" * 60)
print("FINAL PREPROCESSING SUMMARY")
print("=" * 60)

print("Final Shape:", df.shape)

print(
    "Numerical Features:",
    len(numerical_cols)
)

print(
    "Categorical Features:",
    len(categorical_cols)
)

print(
    "Missing Before:",
    f"{missing_before:,}"
)

print(
    "Missing After:",
    f"{missing_after:,}"
)

print("=" * 60)

ORIGINAL DATASET
Shape: (9082, 3925)

After Removing Unnamed: 0
Shape: (9082, 3924)

Columns >80% Missing:
908

After Removing High Missing Columns
Shape: (9082, 3016)

Numerical Features: 3007
Categorical Features: 8

Missing Values Before Treatment:
2,255,733

Filling Numerical Features...
Filling Categorical Features...

Missing Values After Treatment:
0

Target Distribution
F3924
0    9001
1      81
Name: count, dtype: int64

Preprocessed Dataset Saved

FINAL PREPROCESSING SUMMARY
Final Shape: (9082, 3016)
Numerical Features: 3007
Categorical Features: 8
Missing Before: 2,255,733
Missing After: 0


Notebook 02 successfully transformed the raw dataset into a machine-learning-ready dataset. Unnecessary columns and highly incomplete features were removed, reducing the feature space from 3925 to 3016 columns. All remaining missing values (2,255,733) were treated using median imputation for numerical features and "Unknown" imputation for categorical features, resulting in a fully complete dataset with zero missing values. This preprocessed dataset is now ready for feature exploration, feature selection, and model development.